File: service/app/src/test/java/io/miragon/blueprint/application/service/RequestOrderCancellationServiceTest.java
package io.miragon.blueprint.application.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import io.miragon.blueprint.application.port.outbound.BikeDealerPort;
import io.miragon.blueprint.domain.bike.OrderId;
import org.junit.jupiter.api.Test;

class RequestOrderCancellationServiceTest {

    private final BikeDealerPort bikeDealer = mock(BikeDealerPort.class);
    private final RequestOrderCancellationService underTest = new RequestOrderCancellationService(bikeDealer);

    @Test
    void requestCancellationAsksTheDealerAndReturnsItsAnswer() {
        // given: a placed order the dealer allows cancelling
        OrderId orderId = new OrderId("ORDER-900");
        when(bikeDealer.requestCancellation(orderId)).thenReturn(true);
        // when: the dealer is asked whether it can be cancelled
        boolean possible = underTest.requestCancellation(orderId);
        // then: the dealer's answer is returned
        assertThat(possible).isTrue();
        verify(bikeDealer).requestCancellation(orderId);
        verifyNoMoreInteractions(bikeDealer);
    }

    @Test
    void requestCancellationReturnsFalseWhenTheDealerRefuses() {
        // given: an order the dealer will not allow cancelling
        OrderId orderId = new OrderId("ORDER-901");
        when(bikeDealer.requestCancellation(orderId)).thenReturn(false);
        // when: the dealer is asked whether it can be cancelled
        boolean possible = underTest.requestCancellation(orderId);
        // then: the dealer's negative answer is propagated unchanged
        assertThat(possible).isFalse();
        verify(bikeDealer).requestCancellation(orderId);
        verifyNoMoreInteractions(bikeDealer);
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/application/service/RequestOrderCancellationServiceTest.kt
package io.miragon.blueprint.application.service

import io.miragon.blueprint.application.port.outbound.BikeDealerPort
import io.miragon.blueprint.domain.bike.OrderId
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class RequestOrderCancellationServiceTest {

    private val bikeDealer = mockk<BikeDealerPort>()
    private val underTest = RequestOrderCancellationService(bikeDealer = bikeDealer)

    @Test
    fun `requestCancellation asks the dealer and returns its answer`() {
        // given: a placed order the dealer allows cancelling
        val orderId = OrderId("ORDER-900")
        every { bikeDealer.requestCancellation(orderId) } returns true
        // when: the dealer is asked whether it can be cancelled
        val possible = underTest.requestCancellation(orderId)
        // then: the dealer's answer is returned
        assertThat(possible).isTrue()
        verify { bikeDealer.requestCancellation(orderId) }
        confirmVerified(bikeDealer)
    }
}
